File: framework/languages/dotnet/e2e/SpotService/Client/Scenarios/SmD4Scenario.cs
using SpotService.Client;
using SpotService.Shared;
using Systems.Zlink.Stream.Connector.Contracts;

namespace SpotService.Client.Scenarios;

internal static class SmD4Scenario
{
    public static async Task RunAsync(string sessionAStreamEndpoint)
    {
        await using var client = await ConnectAndBindWithRetryAsync(sessionAStreamEndpoint);

        var x = await client.Request(new ActorPingReq("to-x"))
            .PacketName("ActorPingReq")
            .Metadata(SpotServiceNames.ActorIdMetadata, "actor-sm-d4-x")
            .Async<ActorPingReply>();
        var y = await client.Request(new ActorPingReq("to-y"))
            .PacketName("ActorPingReq")
            .Metadata(SpotServiceNames.ActorIdMetadata, "actor-sm-d4-y")
            .Async<ActorPingReply>();
        ScenarioAssert.That(x.ActorId == "actor-sm-d4-x" && x.Value == "to-x", "SM-D4 x relay mismatch.");
        ScenarioAssert.That(y.ActorId == "actor-sm-d4-y" && y.Value == "to-y", "SM-D4 y relay mismatch.");

        var xPushed = client.WaitFor<ActorPushNotify>()
            .Where(message => message.Payload.ActorId == "actor-sm-d4-x")
            .Async().AsTask();
        var xPushReply = await client.Request(new ActorPushReq("push-x"))
            .PacketName("ActorPushReq")
            .Metadata(SpotServiceNames.ActorIdMetadata, "actor-sm-d4-x")
            .Async<ActorPingReply>();
        var xNotify = await xPushed;
        ScenarioAssert.That(xPushReply.ActorId == "actor-sm-d4-x", "SM-D4 x push reply actor mismatch.");
        ScenarioAssert.That(xNotify.Payload.Value == "push-x", "SM-D4 x push payload mismatch.");

        var yPushed = client.WaitFor<ActorPushNotify>()
            .Where(message => message.Payload.ActorId == "actor-sm-d4-y")
            .Async().AsTask();
        var yPushReply = await client.Request(new ActorPushReq("push-y"))
            .PacketName("ActorPushReq")
            .Metadata(SpotServiceNames.ActorIdMetadata, "actor-sm-d4-y")
            .Async<ActorPingReply>();
        var yNotify = await yPushed;
        ScenarioAssert.That(yPushReply.ActorId == "actor-sm-d4-y", "SM-D4 y push reply actor mismatch.");
        ScenarioAssert.That(yNotify.Payload.Value == "push-y", "SM-D4 y push payload mismatch.");

        await ExpectFailureAsync(
            client.Request(new ActorPingReq("missing-actor-id"))
                .PacketName("ActorPingReq")
                .Timeout(TimeSpan.FromSeconds(2))
                .Async<ActorPingReply>().AsTask(),
            "SM-D4 expected actor-id-less request to fail with multiple bound actors.");

        Console.WriteLine("operation SpotService.sm-d4 passed");
    }

    static async Task<IZlinkStreamConnector> ConnectAndBindWithRetryAsync(string endpoint)
    {
        var deadline = DateTimeOffset.UtcNow + TimeSpan.FromSeconds(10);
        Exception? last = null;
        while (DateTimeOffset.UtcNow < deadline)
        {
            var client = CreateClient(endpoint);
            try
            {
                await client.Connect.Async();
                var bound = await client.Request(new MultiBindReq("actor-sm-d4-x", "actor-sm-d4-y", "play-a"))
                    .PacketName("MultiBindReq")
                    .Async<MultiBindReply>();
                ScenarioAssert.That(bound.BoundCount == 2, "SM-D4 expected two bound actors.");
                return client;
            }
            catch (Exception ex) when (ex is ZlinkStreamException or TimeoutException)
            {
                last = ex;
                await client.DisposeAsync();
                await Task.Delay(200);
            }
        }

        throw new InvalidOperationException(
            last is null ? "SM-D4 multi-bind did not become routable." : $"SM-D4 multi-bind did not become routable. Last error: {last.Message}",
            last);
    }

    static IZlinkStreamConnector CreateClient(string endpoint)
    {
        ScenarioAssert.That(!string.IsNullOrWhiteSpace(endpoint), "session-a stream endpoint is required.");
        return ZlinkStreamConnectorFactory.Create(new ZlinkStreamConnectorOptions
        {
            Endpoint = new Uri(endpoint),
            ConnectTimeout = TimeSpan.FromSeconds(5),
            RequestTimeout = TimeSpan.FromSeconds(5),
            Heartbeat = new ZlinkStreamHeartbeatOptions { Enabled = false },
            DispatchMode = ZlinkStreamDispatchMode.Immediate,
            MaxReceivedMessages = 1024,
        });
    }

    static async Task ExpectFailureAsync(Task task, string message)
    {
        try
        {
            await task;
        }
        catch
        {
            return;
        }

        throw new InvalidOperationException(message);
    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-d4-scenario.ts
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode
} from '@zlink-systems/stream-connector';
import type {
  ActorPingRes,
  ActorPingReq,
  ActorPushNotify,
  ActorPushReq,
  EvidenceWaitReq,
  MultiBindRes,
  MultiBindReq
} from '../../Shared/messages';
import { SpotServiceNames } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../Support/http-client';
import { ensure } from '../Support/scenario-assert';
import { decodeStreamReply } from '../Support/stream-reply';

export async function runSmD4(options: ClientOptions): Promise<void> {
  const firstActorId = 'actor-sm-d4-x';
  const secondActorId = 'actor-sm-d4-y';
  const client = zlinkStreamConnectorFactory.create({
    endpoint: options.sessionAStreamEndpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    waitTimeoutMs: 10000
  });
  await client.connect();
  try {
    const bound = decodeStreamReply<MultiBindRes>(await client
      .request({
        firstActorId,
        secondActorId,
        nodeRid: 'play-a'
      } satisfies MultiBindReq)
      .packetName('MultiBindReq')
      .timeout(5000)
      .submit());
    ensure(bound.boundCount === 2, 'SM-D4 expected two bound actors.');

    const x = decodeStreamReply<ActorPingRes>(await client
      .request({ value: 'to-x' } satisfies ActorPingReq)
      .packetName('ActorPingReq')
      .metadata(SpotServiceNames.actorIdMetadata, firstActorId)
      .timeout(5000)
      .submit());
    const y = decodeStreamReply<ActorPingRes>(await client
      .request({ value: 'to-y' } satisfies ActorPingReq)
      .packetName('ActorPingReq')
      .metadata(SpotServiceNames.actorIdMetadata, secondActorId)
      .timeout(5000)
      .submit());
    ensure(x.actorId === firstActorId && x.value === 'to-x', 'SM-D4 x relay mismatch.');
    ensure(y.actorId === secondActorId && y.value === 'to-y', 'SM-D4 y relay mismatch.');

    const xPushed = client.waitFor<ActorPushNotify>('ActorPushNotify')
      .where((message) => message.payload.actorId === firstActorId)
      .timeout(10000)
      .submit();
    const xPushReply = decodeStreamReply<ActorPingRes>(await client
      .request({ value: 'push-x' } satisfies ActorPushReq)
      .packetName('ActorPushReq')
      .metadata(SpotServiceNames.actorIdMetadata, firstActorId)
      .timeout(5000)
      .submit());
    const xNotify = await xPushed;
    ensure(xPushReply.actorId === firstActorId, 'SM-D4 x push reply actor mismatch.');
    ensure(xNotify.payload.value === 'push-x', 'SM-D4 x push payload mismatch.');

    const yPushed = client.waitFor<ActorPushNotify>('ActorPushNotify')
      .where((message) => message.payload.actorId === secondActorId)
      .timeout(10000)
      .submit();
    const yPushReply = decodeStreamReply<ActorPingRes>(await client
      .request({ value: 'push-y' } satisfies ActorPushReq)
      .packetName('ActorPushReq')
      .metadata(SpotServiceNames.actorIdMetadata, secondActorId)
      .timeout(5000)
      .submit());
    const yNotify = await yPushed;
    ensure(yPushReply.actorId === secondActorId, 'SM-D4 y push reply actor mismatch.');
    ensure(yNotify.payload.value === 'push-y', 'SM-D4 y push payload mismatch.');

    let actorIdLessRequestFailed = false;
    try {
      await client
        .request({ value: 'missing-actor-id' } satisfies ActorPingReq)
        .packetName('ActorPingReq')
        .timeout(2000)
        .submit();
    } catch {
      actorIdLessRequestFailed = true;
    }
    ensure(actorIdLessRequestFailed, 'SM-D4 expected actor-id-less request to fail with multiple bound actors.');
  } finally {
    await client.close();
  }
  await postJson<string[]>(options.playAUrl, '/evidence/wait', {
    containsAll: [`entry-disconnected|rid=play-a|actor=${firstActorId}`],
    timeoutMilliseconds: 10000
  } satisfies EvidenceWaitReq);

  console.log('scenario SM-D4 passed');
}
